App crashes on launch

App crashes on every launch (except the first launch)
```FATAL EXCEPTION: main
                                                                         Process: in.tosc.digitaloceanapp, PID: 18707
                                                                         java.lang.RuntimeException: Unable to start activity ComponentInfo{in.tosc.digitaloceanapp/in.tosc.digitaloceanapp.DropletActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.Object retrofit2.Retrofit.create(java.lang.Class)' on a null object reference
                                                                             at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2456)
                                                                             at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2539)
                                                                             at android.app.ActivityThread.access$900(ActivityThread.java:159)
                                                                             at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1384)
                                                                             at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                             at android.os.Looper.loop(Looper.java:152)
                                                                             at android.app.ActivityThread.main(ActivityThread.java:5507)
                                                                             at java.lang.reflect.Method.invoke(Native Method)
                                                                             at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                                             at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
                                                                             at de.robv.android.xposed.XposedBridge.main(XposedBridge.java:102)
                                                                          Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.Object retrofit2.Retrofit.create(java.lang.Class)' on a null object reference
                                                                             at in.tosc.doandroidlib.DigitalOcean.getDOClient(DigitalOcean.java:109)
                                                                             at in.tosc.digitaloceanapp.DropletActivity.onCreate(DropletActivity.java:65)
                                                                             at android.app.Activity.performCreate(Activity.java:6304)
                                                                             at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1118)
                                                                             at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2409)
                                                                             at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2539) 
                                                                             at android.app.ActivityThread.access$900(ActivityThread.java:159) 
                                                                             at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1384) 
                                                                             at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                             at android.os.Looper.loop(Looper.java:152) 
                                                                             at android.app.ActivityThread.main(ActivityThread.java:5507) 
                                                                             at java.lang.reflect.Method.invoke(Native Method) 
                                                                             at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
                                                                             at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 
                                                                             at de.robv.android.xposed.XposedBridge.main(XposedBridge.java:102) ```

--- a/app/src/main/java/in/tosc/digitaloceanapp/activities/DetailDropletActivity.java
+++ b/app/src/main/java/in/tosc/digitaloceanapp/activities/DetailDropletActivity.java
@@ -49,7 +49,7 @@ public class DetailDropletActivity extends AppCompatActivity implements Compound
         Gson gson = new Gson();
         droplet = gson.fromJson(getIntent().getStringExtra("DROPLET"),Droplet.class);
 
-        doaClient = DigitalOcean.getDOClient();
+        doaClient = DigitalOcean.getDOClient(getSharedPreferences("DO", MODE_PRIVATE).getString("authToken",null));
 
         coordinatorLayout = (CoordinatorLayout) findViewById(R.id.coordinator_layout);
 

--- a/app/src/main/java/in/tosc/digitaloceanapp/activities/DropletActivity.java
+++ b/app/src/main/java/in/tosc/digitaloceanapp/activities/DropletActivity.java
@@ -55,7 +55,6 @@ public class DropletActivity extends AppCompatActivity
         super.onCreate(savedInstanceState);
 
         FontsOverride.applyFontForToolbarTitle(this, FontsOverride.FONT_PROXIMA_NOVA);
-
         setContentView(R.layout.activity_droplet);
         Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
         setSupportActionBar(toolbar);
@@ -63,7 +62,7 @@ public class DropletActivity extends AppCompatActivity
         dropletRecyclerView = (RecyclerView) findViewById(R.id.dropletsRv);
         dropletRecyclerView.setLayoutManager(new LinearLayoutManager(DropletActivity.this, LinearLayoutManager.VERTICAL, false));
         dropletRecyclerView.setAdapter(dropletsAdapter);
-        DigitalOceanClient doClient = DigitalOcean.getDOClient();
+        DigitalOceanClient doClient = DigitalOcean.getDOClient(getSharedPreferences("DO", MODE_PRIVATE).getString("authToken",null));
         refreshData();
 
         swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
@@ -107,7 +106,7 @@ public class DropletActivity extends AppCompatActivity
 
 
     private void refreshData() {
-        DigitalOceanClient doClient = DigitalOcean.getDOClient();
+        DigitalOceanClient doClient = DigitalOcean.getDOClient(getSharedPreferences("DO", MODE_PRIVATE).getString("authToken",null));
         doClient.getDroplets(1,10).enqueue(new Callback<List<Droplet>>() {
             @Override
             public void onResponse(Call<List<Droplet>> call, Response<List<Droplet>> response) {

--- a/app/src/main/java/in/tosc/digitaloceanapp/activities/SampleActivity.java
+++ b/app/src/main/java/in/tosc/digitaloceanapp/activities/SampleActivity.java
@@ -28,8 +28,8 @@ public class SampleActivity extends AppCompatActivity {
 
         FontsOverride.applyFontForToolbarTitle(this, "ProximaNova.ttf");
 
-        DigitalOceanClient doClient = DigitalOcean.getDOClient();
-        final DigitalOceanStatisticsClient statClient = DigitalOcean.getDOStatsClient();
+        DigitalOceanClient doClient = DigitalOcean.getDOClient(getSharedPreferences("DO", MODE_PRIVATE).getString("authToken",null));
+        final DigitalOceanStatisticsClient statClient = DigitalOcean.getDOStatsClient(getSharedPreferences("DO", MODE_PRIVATE).getString("authToken",null));
 
 
 

--- a/app/src/main/java/in/tosc/digitaloceanapp/fragments/SelectImageFragment.java
+++ b/app/src/main/java/in/tosc/digitaloceanapp/fragments/SelectImageFragment.java
@@ -20,6 +20,8 @@ import retrofit2.Call;
 import retrofit2.Callback;
 import retrofit2.Response;
 
+import static android.content.Context.MODE_PRIVATE;
+
 public class SelectImageFragment extends Fragment {
 
     List<Image> imageList;
@@ -38,7 +40,7 @@ public class SelectImageFragment extends Fragment {
     public View onCreateView(LayoutInflater inflater, ViewGroup container,
                              Bundle savedInstanceState) {
         View view = inflater.inflate(R.layout.fragment_image, container, false);
-        DigitalOceanClient doClient = DigitalOcean.getDOClient();
+        DigitalOceanClient doClient = DigitalOcean.getDOClient(getContext().getSharedPreferences("DO", MODE_PRIVATE).getString("authToken",null));
         imageRecyclerView = (RecyclerView) view.findViewById(R.id.imageRecyclerVIew);
         imageRecyclerView.setLayoutManager(new GridLayoutManager(getContext(),3));
         imageRecyclerView.setAdapter(imageAdapter);

--- a/app/src/main/java/in/tosc/digitaloceanapp/fragments/SelectSizeFragment.java
+++ b/app/src/main/java/in/tosc/digitaloceanapp/fragments/SelectSizeFragment.java
@@ -3,7 +3,6 @@ package in.tosc.digitaloceanapp.fragments;
 import android.os.Bundle;
 import android.support.annotation.Nullable;
 import android.support.v4.app.Fragment;
-import android.support.v4.app.ListFragment;
 import android.support.v7.widget.GridLayoutManager;
 import android.support.v7.widget.RecyclerView;
 import android.util.Log;
@@ -14,15 +13,15 @@ import android.view.ViewGroup;
 import java.util.List;
 
 import in.tosc.digitaloceanapp.R;
-import in.tosc.digitaloceanapp.adapters.ImageAdapter;
 import in.tosc.digitaloceanapp.adapters.SelectSizeAdapter;
 import in.tosc.doandroidlib.DigitalOcean;
 import in.tosc.doandroidlib.api.DigitalOceanClient;
-import in.tosc.doandroidlib.objects.Image;
 import in.tosc.doandroidlib.objects.Size;
 import retrofit2.Call;
 import retrofit2.Callback;
 import retrofit2.Response;
+
+import static android.content.Context.MODE_PRIVATE;
 
 /**
  * Created by the-dagger on 11/26/2016.
@@ -39,7 +38,7 @@ public class SelectSizeFragment extends Fragment{
     public View onCreateView(LayoutInflater inflater, @Nullable final ViewGroup container, @Nullable Bundle savedInstanceState) {
         View view = inflater.inflate(R.layout.fragment_size, container, false);
 
-        DigitalOceanClient doClient = DigitalOcean.getDOClient();
+        DigitalOceanClient doClient = DigitalOcean.getDOClient(getContext().getSharedPreferences("DO", MODE_PRIVATE).getString("authToken",null));
         recyclerView = (RecyclerView) view.findViewById(R.id.recyclerView_size);
         recyclerView.setLayoutManager(new GridLayoutManager(getContext(),1));
         recyclerView.setAdapter(selectSizeAdapter);

--- a/doandroidlib/src/main/java/in/tosc/doandroidlib/DigitalOcean.java
+++ b/doandroidlib/src/main/java/in/tosc/doandroidlib/DigitalOcean.java
@@ -155,14 +155,6 @@ public class DigitalOcean {
                     .build();
 
 
-            Gson gson = new GsonBuilder()
-                    .registerTypeAdapter(Account.class, new ObjectDeserializer<Account>("account"))
-                    .registerTypeAdapter(Action.class, new ObjectDeserializer<Action>("action"))
-                    .registerTypeAdapter(new TypeToken<List<Droplet>>(){}.getType(), new ListDeserializer<Droplet>("droplets"))
-                    .registerTypeAdapter(new TypeToken<List<Image>>(){}.getType(), new ListDeserializer<Image>("images"))
-                    .registerTypeAdapter(new TypeToken<List<Size>>(){}.getType(), new ListDeserializer<Size>("sizes"))
-                    .create();
-
             r2 = new Retrofit.Builder()
                     .baseUrl(BASE_URL_V1)
                     .addConverterFactory(GsonConverterFactory.create())

--- a/doandroidlib/src/main/java/in/tosc/doandroidlib/DigitalOcean.java
+++ b/doandroidlib/src/main/java/in/tosc/doandroidlib/DigitalOcean.java
@@ -105,11 +105,70 @@ public class DigitalOcean {
                 .build();
     }
 
-    public static DigitalOceanClient getDOClient () {
+    public static DigitalOceanClient getDOClient (final String authToken) {
+
+        if(r == null){
+            OkHttpClient httpClient = new OkHttpClient.Builder()
+                    .addInterceptor(new Interceptor() {
+                        @Override
+                        public Response intercept(Chain chain) throws IOException {
+                            Request request = chain.request().newBuilder()
+                                    .addHeader("Content-Type", "application/json")
+                                    .addHeader("Authorization", "Bearer " + authToken)
+                                    .build();
+                            return chain.proceed(request);
+                        }
+                    })
+                    .build();
+
+
+            Gson gson = new GsonBuilder()
+                    .registerTypeAdapter(Account.class, new ObjectDeserializer<Account>("account"))
+                    .registerTypeAdapter(Action.class, new ObjectDeserializer<Action>("action"))
+                    .registerTypeAdapter(new TypeToken<List<Droplet>>(){}.getType(), new ListDeserializer<Droplet>("droplets"))
+                    .registerTypeAdapter(new TypeToken<List<Image>>(){}.getType(), new ListDeserializer<Image>("images"))
+                    .registerTypeAdapter(new TypeToken<List<Size>>(){}.getType(), new ListDeserializer<Size>("sizes"))
+                    .create();
+
+            r = new Retrofit.Builder()
+                    .baseUrl(BASE_URL)
+                    .addConverterFactory(GsonConverterFactory.create(gson))
+                    .client(httpClient)
+                    .build();
+        }
         return r.create(DigitalOceanClient.class);
     }
 
-    public static DigitalOceanStatisticsClient getDOStatsClient () {
+    public static DigitalOceanStatisticsClient getDOStatsClient (final String authToken) {
+        if(r2 == null){
+            OkHttpClient httpClient = new OkHttpClient.Builder()
+                    .addInterceptor(new Interceptor() {
+                        @Override
+                        public Response intercept(Chain chain) throws IOException {
+                            Request request = chain.request().newBuilder()
+                                    .addHeader("Content-Type", "application/json")
+                                    .addHeader("Authorization", "Bearer " + authToken)
+                                    .build();
+                            return chain.proceed(request);
+                        }
+                    })
+                    .build();
+
+
+            Gson gson = new GsonBuilder()
+                    .registerTypeAdapter(Account.class, new ObjectDeserializer<Account>("account"))
+                    .registerTypeAdapter(Action.class, new ObjectDeserializer<Action>("action"))
+                    .registerTypeAdapter(new TypeToken<List<Droplet>>(){}.getType(), new ListDeserializer<Droplet>("droplets"))
+                    .registerTypeAdapter(new TypeToken<List<Image>>(){}.getType(), new ListDeserializer<Image>("images"))
+                    .registerTypeAdapter(new TypeToken<List<Size>>(){}.getType(), new ListDeserializer<Size>("sizes"))
+                    .create();
+
+            r2 = new Retrofit.Builder()
+                    .baseUrl(BASE_URL_V1)
+                    .addConverterFactory(GsonConverterFactory.create())
+                    .client(httpClient)
+                    .build();
+        }
         return r2.create(DigitalOceanStatisticsClient.class);
     }
 

